Change dataloader send_to_device calls to non-blocking

drhead opened the pull request:
# What does this PR do?

Currently, it seems that even with pinned memory enabled on the dataloader, iterating over the dataloader causes a cudaStreamSynchronize call:

![image](https://github.com/huggingface/accelerate/assets/1313496/1e356748-49fb-46e0-b812-fec77e598ef1)

This is a devastating hit to performance if you are trying to overlap any other asynchronous transfers with your computations (such as offloading optimizer/EMA states), since it appears to be waiting for all async transfers to complete before continuing.

Setting these operations to be non-blocking, as well as using pinned memory for the dataloader, gets rid of the offending call:
![image](https://github.com/huggingface/accelerate/assets/1313496/4876b3cb-de20-46f4-ac6d-b7cadf6e1507)
(the timescale is very different here, this one is a fraction of a millisecond whereas the previous screenshot highlights a period of roughly one entire second)

With this, the tensors should only get forcibly synced when they are about to be used, and other non-blocking transfers will be left alone.  When I made this change, I had a noticeable speedup in my training loop where I am offloading an EMA state (3s/it -> 2.5s/it).

This also shouldn't cause issues when not using pinned memory.  It will just do a normal synchronous transfer as before.  The `send_to_device` utility function also has handling built in for any case where `Tensor.to()` does not accept `non_blocking` as an argument for whatever reason.  There shouldn't be any downsides to this change, as a result, so I don't see any reason to leave it as anything but a fixed value of True.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?


## Who can review?

- Core parts of the library: @muellerzr @BenjaminBossan

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/accelerate/pr_2685). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

muellerzr commented:
As stated back here, we have reason to keep it at pytorch's default of `False`. What we can do though is add `non_blocking` as an argument to `utils.dataclasses.DataLoaderConfiguration` and have the user set it there. https://github.com/huggingface/accelerate/pull/607

Does this seem reasonable for you? :)

drhead commented:
> As stated back here, we have reason to keep it at pytorch's default of `False`. What we can do though is add `non_blocking` as an argument to `utils.dataclasses.DataLoaderConfiguration` and have the user set it there.
> 
> Does this seem reasonable for you? :)

I do think that would work.  I will look into setting that up and create appropriate documentation for it, including a warning about CPU operations on non-blocking tensors.  I do think that the common use case for a dataloader isn't threatened by this since this will usually be a single HtoD transfer with the tensor destroyed on device.

drhead commented:
This implementation works and has been tested.  I did add documentation notes where appropriate on the classes I modified and also added a warning in the Quickstart section about how to enable non-blocking transfers when migrating, it seems appropriate to bring it up there since that is where non-blocking calls would be made in vanilla Pytorch code.

muellerzr commented on src/accelerate/utils/dataclasses.py:
```diff
@@ -524,6 +524,14 @@ class DataLoaderConfiguration:
+        default=False,
+        metadata={
+            "help": "If set to `True`, the dataloader prepared by the Accelerator will utilize non-blocking host-to-device"
+            " transfers, allowing for better overlap between dataloader communication and computation.  Requires that the"
+            " prepared dataloader has `pin_memory` set to `True` to work properly."
```
```suggestion
            " transfers, allowing for better overlap between dataloader communication and computation.  Recommended that the"
            " prepared dataloader has `pin_memory` set to `True` to work properly."
```

Otherwise, we should raise a relevant error :)

muellerzr reviewed (approved):
Very nice! Just one small nit and we're good to go.

Also, you can fix the quality issues by doing:

```bash
pip install -e .[quality]
make style; make quality
```

muellerzr reviewed (approved):
Thanks a bunch! Nice work 🤗
